### PR TITLE
Attribute Event purging for user changes

### LINKED ISSUE #3336: Unnecessary purges on `User` model change
Currently right now any time the `User` model changes, we purge all the cache keys for that user's project.

This includes attribute changes that don't actually affect the project pages, like `last_login`, `password` etc.

We should filter out "purge-able" attribute changes and only issue purges when necessary. Said attributes include:
* `username`
* `name`
* `emails`

### PR DESCRIPTION
**Pull request by di**
Fixes #3336.

### CODE REVIEW

**Comment by dstufft** [comment]
I just woke up, but this seems to simply be adding more cases where things will get purged, but to actually fix #3336 we need to remove cases?

**Comment by di** [comment]
@dstufft The diff is tricky but this removes all the `purge_keys` for the `User` model:

```diff
config.register_origin_cache_keys(
    User,
    cache_keys=["user/{obj.username}"],
-     purge_keys=[
-         key_factory("user/{obj.username}"),
-         key_factory(
-             "project/{itr.normalized_name}",
-             iterate_on='projects',
-         ),
-     ],
- ),
```

And adds them back only on specific attributes:

```diff
+    config.register_origin_cache_keys(
+        User.name,
+        purge_keys=[
+            key_factory("user/{obj.user.username}"),
+            key_factory(
+                "project/{itr.normalized_name}",
+                iterate_on='user.projects',
+            )
+        ],
+    )
+    config.register_origin_cache_keys(
+        Email.primary,
+        purge_keys=[
+            key_factory("user/{obj.user.username}"),
+            key_factory(
+                "project/{itr.normalized_name}",
+                iterate_on='user.projects',
+            )
+        ],
+    )
```

**Comment by dstufft** [comment]
@di did you forget to push something? That's not what I see in the changes tab.

**Comment by dstufft** [comment]
Oh wait I see it, fucking git.

File: warehouse/packaging/__init__.py [hunk 1]
```diff
@@ -49,11 +60,24 @@ def includeme(config):
     config.register_origin_cache_keys(
         User,
         cache_keys=["user/{obj.username}"],
+    )
+    config.register_origin_cache_keys(
+        User.name,
```
Comment (di) [inline warehouse/packaging/__init__.py]: @dstufft Adding these lines are removing the purge keys from `User` and adding the same ones for `User.name`. Sorry I didn't split this into two commits, that may have been more clear.

**Comment by di** [comment]
@ewdurbin On it.

**Pull request merged**